repository_service_tuf_worker/services/storage/awss3.py

```diff
@@ -0,0 +1,131 @@
+from typing import List, Optional
+
+import boto3
+from securesystemslib.exceptions import StorageError  # noqa
+from tuf.api.metadata import Metadata, T, Timestamp
+from tuf.api.serialization import DeserializationError
+
+from repository_service_tuf_worker.interfaces import IStorage, ServiceSettings
+
+
+class AWSS3(IStorage):
+    def __init__(
+        self, bucket: str, access_key: str, secret_key: str, region: str
+    ) -> None:
+        self._bucket: str = bucket
+        self._access_key: str = access_key
+        self._secret_key: str = secret_key
+        self._region: str = region
+
+        self._s3 = boto3.client(
+            "s3",
+            aws_access_key_id=self._access_key,
+            aws_secret_access_key=self._secret_key,
+            region_name=self._region,
+            endpoint_url="http://localstack:4566",
+        )
+
+        self._s3_resource = boto3.resource(
+            "s3",
+            aws_access_key_id=self._access_key,
+            aws_secret_access_key=self._secret_key,
+            region_name=self._region,
+            endpoint_url="http://localstack:4566",
+        )
+
+    @classmethod
+    def configure(cls, settings) -> None:
+        s3_resource = boto3.resource(
+            "s3",
+            aws_access_key_id=settings.AWSS3_STORAGE_ACCESS_KEY,
+            aws_secret_access_key=settings.AWSS3_STORAGE_SECRET_KEY,
+            region_name=settings.AWSS3_STORAGE_REGION,
+            endpoint_url="http://localstack:4566",
+        )
+
+        buckets = [bucket.name for bucket in s3_resource.buckets.all()]
+        if settings.AWSS3_STORAGE_BUCKET not in buckets:
+            raise ValueError(
+                f"Bucket {settings.AWSS3_STORAGE_BUCKET} not found."
+            )
+
+    @classmethod
+    def settings(cls) -> List[ServiceSettings]:
+        return [
+            ServiceSettings(
+                name="AWSS3_STORAGE_BUCKET",
+                argument="bucket",
+                required=True,
+            ),
+            ServiceSettings(
+                name="AWSS3_STORAGE_ACCESS_KEY",
+                argument="access_key",
+                required=True,
+            ),
+            ServiceSettings(
+                name="AWSS3_STORAGE_SECRET_KEY",
+                argument="secret_key",
+                required=True,
+            ),
+            ServiceSettings(
+                name="AWSS3_STORAGE_REGION",
+                argument="region",
+                required=True,
+            ),
+        ]
+
+    def get(self, role: str, version: Optional[int] = None) -> Metadata[T]:
+        """
+        Returns TUF role metadata object for the passed role name,
+        optionally at the passed version (latest if None).
+        """
+
+        bucket = self._s3_resource.Bucket(self._bucket)
+        if role == Timestamp.type:
+            filename = f"{role}.json"
+        else:
+            if version is None:
+                filenames = [file.key for file in bucket.objects.all()]
```

Comment:
We have a challenge here.

The Specification and TUF Client expects `<version>.<role>.json`
The AWS S3 only has a filter by `prefix`, which doesn't allow us to filter like `*.<role_name>.json` to get all the existent role versions.
Another issue is that the `objects.all()` is limited to 1,000 objects per page, which creates pagination.

I'm looking for solutions.

Reply:
Also: https://howfuckedismydatabase.com/nosql/ 😝

Reply:
> Maybe we can use [`awswrangler`](https://aws-sdk-pandas.readthedocs.io/en/stable/stubs/awswrangler.s3.list_objects.html#awswrangler.s3.list_objects), it seems to support wildcards? Alternatively, we could store metadata, e.g. as `<rolename>/<version>`. AFAIU results are returned in UTF-8 binary order (ascending?). So maybe we can grab the last item on the last page to get the latest version for a given prefix.

I thought about `awswrangler` as well

Reply:
> Alternatively, we could store metadata, e.g. as `<rolename>/<version>`. AFAIU results are returned in UTF-8 binary order (ascending?). So maybe we can grab the last item on the last page to get the latest version for a given prefix.

What about the clients accessing, they will look for `<version>.<rolename>.json`, right?